Add option for daemon kube-apiserver access to bypass host firewall

atykhyy opened the pull request:
When L7 DNS proxy for nodes is enabled and kube-apiserver is a FQDN, Cilium agent deadlocks on restart, because k8sClientset's start hook which checks connection to kube-apiserver and verifies the version runs before the daemon bootstraps the DNS proxy with restored endpoints. This commit adds a daemon command-line option which configures the dialer of k8sClientset to bypass DNS proxy and host firewall in the same manner as DNS proxy's requests to remote DNS servers do. Since Cilium trusts both DNS and kube-apiserver responses, exempting the daemon's connections to kube-apiserver from host firewall does not degrade security.

Note: both DNS resolution of and actual connection to kube-apiserver are exempted. If only DNS resolution of kube-apiserver's address is exempted, DNS proxy would not create correct IP-based rules for its address, and if kube-apiserver's IP address changes while the Cilium daemon is restarting, kube-apiserver will not be accessible and the daemon will fail on startup.


```release-note
Introduce the option `enable-k8s-host-firewall-bypass` to allow cilium to bypass the host firewall both for resolving (when configured as a FQDN) and connecting to kube-apiserver.  
```

Fixes: #35433

asauber commented on daemon/cmd/datapath.go:
```diff
@@ -364,3 +369,37 @@ func setupRouteToVtepCidr(logger *slog.Logger) error {
+	if err := c.Control(func(su uintptr) {
+		secId := identity.ReservedIdentityHost
+		mark := linux_defaults.MagicMarkEgress
+		mark |= uint32(secId&0xFFFF)<<16 | uint32((secId&0xFF0000)>>16)
```
Since this always results in the same set of bits, derived from constants, can we add this mark as a constant and document how it's derived with a comment?

asauber replied:
That being said, I suppose that the other two constants involved here could change at some point, so maybe what would be most useful would be a comment about this "mask and 16 bit swap", which I guess appears elsewhere in the codebase but could use some explanation here.

atykhyy replied:
I factored this computation out and put it where it is explained.

asauber reviewed (approved):
The justification for this implementation in https://github.com/cilium/cilium/issues/35433 looks reasonable to me. Approving for the configuration implementation. One comment about the readability.

bimmlerd commented on daemon/cmd/cells.go:
```diff
@@ -104,6 +104,10 @@ var (
 
+		// Provides optional configuration callback to enable bypassing
+		// host firewall when accessing Kubernetes objects.
+		cell.Provide(newEnableK8sHostFirewallBypass),
```
nit: this strikes me as somewhat out of place; can this concern be handled internally to the k8s client cell? Also, maybe drop the enable from the name

```suggestion
		cell.Provide(newK8sHostFirewallBypass),
```

atykhyy replied:
I considered handling this inside the k8s client cell, but such datapath-specific details seemed out of place there too; it's not like the DNS proxy cell where engaging with the datapath is the main concern. if I understand the code correctly, the k8s client cell is used in other executables, not only in the daemon. So I ended up adding a generic extensibility point to the k8s client cell and configuring it with DI in the daemon.

> drop the enable from `newEnableK8sHostFirewallBypass`

Sure, I'll rename it.

bimmlerd replied:
Fair enough. I don't have a better suggestion, this isn't bad.

bimmlerd commented on daemon/cmd/daemon_main.go:
```diff
@@ -535,6 +535,9 @@ func InitGlobalFlags(logger *slog.Logger, cmd *cobra.Command, vp *viper.Viper) {
 	flags.Bool(option.EnableHostFirewall, false, "Enable host network policies")
 	option.BindEnv(vp, option.EnableHostFirewall)
 
+	flags.Bool(option.K8sEnableHostFirewallBypass, false, "Enable bypassing host firewall for Kubernetes API server access")
```
nit: similarly; if this was internal to the k8s cell; config could live there without adding more cruft to option/config

atykhyy replied:
Actually I believe this bypass should not be optional at all - simply enabled whenever host firewall is enabled. Then there would not be an option for it, and no extra cruft. However, I did not want to add a new behavior without a way to turn it off. I'm not nearly confident enough in my knowledge of Cilium internals to say it's safe. If this change (decorating k8s client's dialers with the proxy egress packet mark) seems safe enough to reviewers to enable in the main branch, I can remove the option and associated logic.

bimmlerd replied:
I'm also not confident enough for that :upside_down_face: Let's see what other reviewers say.

joestringer replied:
Opt out seems like a reasonable approach, then we could file an issue to remove the flag in a couple of releases once it's been used in the wild for a while.

You could also mark the option as "hidden" so it's not advertised in the command reference and whatnot. If it really comes down to it, there would be the option to disable, but it seems like for the vast majority of users we actually don't expect them to use it and we don't anticipate needing the flag.

atykhyy replied:
I have switched the option's default to `true` in the most recent push, making it in effect opt-out. I'm not sure about marking it hidden, though. Shouldn't it be at least somewhat discoverable in case things go wrong?

joestringer replied:
I'm personally fine with hiding flags when we don't think most users are going to need it. There are already far too many configuration knobs in Cilium, and in many cases adding more flags just gives users more ways to break Cilium. If there's two real use cases where we expect users to explicitly enable or disable the flag, then yes we should make it discoverable and document what the behavior is. If we think there's one right option and there's just a small risk of regression then I'd rather not advertise the option to most users.

In this specific case I think the risk someone disables it without understanding and causes a problem is higher than the risk that someone encounters a problem with the default value of the flag and needs to find the option to address their issue.

joestringer replied:
I don't feel super strongly about this either, but it's also one of these things where if we're already thinking about when to get rid of the flag, then we probably shouldn't be trying to teach users to configure it.

atykhyy replied:
Marked the option as hidden.

bimmlerd reviewed (changes requested):
I'm OK with this approach, just some implementation nits

cc @joamaki and @joestringer as you've interacted with the issue this PR fixes

atykhyy committed:
Add option for daemon kube-apiserver access to bypass host firewall

When L7 DNS proxy for nodes is enabled and kube-apiserver is a FQDN,
Cilium agent deadlocks on restart, because k8sClientset's start hook
which checks connection to kube-apiserver and verifies its version runs
before the daemon bootstraps the DNS proxy with restored endpoints.
(The build-config init container is also affected as it by definition
executes while the daemon is not running.)

This commit configures the dialer of daemon and cilium-dbg build-config's
k8sClientset to bypass DNS proxy and host firewall in the same manner
as DNS proxy's requests to remote DNS servers do. The bypass is enabled
by default, but can be disabled with a command-line option. Since Cilium
trusts both DNS and kube-apiserver responses, exempting the daemon's
connections to kube-apiserver from host firewall does not degrade security.

Note: both DNS resolution of and actual connection to kube-apiserver
are exempted. If only DNS resolution of kube-apiserver's address is
exempted, DNS proxy would not create correct IP-based rules for its
address, and if kube-apiserver's IP address changes while the Cilium
daemon is restarting, kube-apiserver will not be accessible and the
daemon will fail on startup.

Fixes: #35433

Signed-off-by: Anton Tykhyy <atykhyy@gmail.com>

squeed commented on pkg/k8s/hostfirewallbypass/bypass.go:
```diff
@@ -0,0 +1,47 @@
+func setProxyEgressMark(network, address string, c syscall.RawConn) error {
+	var soerr error
+	if err := c.Control(func(su uintptr) {
+		mark := linux_defaults.MakeMagicMark(linux_defaults.MagicMarkEgress, identity.ReservedIdentityHost)
```
Hmm, interesting. If I'm reading this correctly, it is marking the packets as having a source identity of the local host. Shouldn't they already have that?

I trust you that it works, but _why_ does it?

atykhyy replied:
> [I]t is marking the packets as having a source identity of the local host
> Shouldn't they already have that?

I believe so, but that labeling happens automatically based on where packets originate. This line is doing more than that: it is marking the packets so that they bypass host firewall in the same manner as the packets DNS proxy sends to remote DNS servers. This is done to avoid a potential deadlock: the daemon (and build-config) need to access kube-apiserver, but this happens earlier in the daemon's startup sequence than DNS proxy initialization, so if there are host L7 DNS rules in effect and kube-apiserver is an FQDN (e.g. in cloud-based kubernetes clusters), k8s client cannot resolve kube-apiserver address and daemon fails on restart. The config init container runs before the agent container, so DNS proxy is not available to it either.

squeed replied:
@jrajahalme so, MagicMarkEgress has the effect of skipping policy. Is this intended or a side-effect?

atykhyy replied:
The comment in pkg/datapath/linux/linux_defaults/mark.go says that
```go
// MagicMarkEgress determines that the traffic is sourced from the
// proxy which is applying Egress policy
```
which makes me believe that it is intended to be applied to packets to which policy has already been applied in a different part of the stack, such as DNS proxy. In this PR's case, the (implicit) policy is "Cilium agent is always allowed to talk to kube-apiserver".

joestringer replied:
> MagicMarkEgress has the effect of skipping policy. Is this intended or a side-effect?

`MagicMarkEgress` is set by the proxy on packets being transmitted out from the proxy. The implication is that the corresponding messages already passed through BPF, hit a policy entry that redirects to the proxy, went to the proxy, which then enforces L7 policy, then hands the message on to the outbound socket. So yes this is expected. If you search this symbol in the tree you'll find similar usage in the dnsproxy.

tommyp1ckles commented on pkg/option/config.go:
```diff
@@ -1754,6 +1757,9 @@ type DaemonConfig struct {
 	EnableHostFirewall bool
 
+	// EnableK8sHostFirewallBypass enables bypassing host firewall for kubernetes api server access
+	EnableK8sHostFirewallBypass bool
```
Could we put this in a `cell.Config` instead of adding to the old global DaemonConfig params? 

For example, with the cell definition for `newK8sHostFirewallBypass` we can do:

```go
                cell.Config(defaultK8sHostFirewallBypass),
		// Provides optional configuration callback to bypass
		// host firewall when accessing Kubernetes objects.
		cell.Provide(newK8sHostFirewallBypass),
```
 
And declare our config as follows:

```go

type k8sHostFirewallBypassParams struct {
                EnableHostFirewallBypass bool
}

var defaultK8sHostFirewallBypass = k8sHostFirewallBypassParams{
		EnableHostFirewallBypass: true
}

func (p k8sHostFirewallBypassParams) Flags(flags *pflag.FlagSet) {
        flags.Bool("enable-k8s-host-firewall-bypass", p.EnableHostFirewallBypass, ""Enable bypassing host firewall for Kubernetes API server access.")
}
```

tommyp1ckles replied:
For the `option.Config.EnableK8sHostFirewallBypass`, we could still rely on the `*option.DaemonConfig` being passed in through hive.

tommyp1ckles commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -154,6 +154,7 @@ cilium-agent [flags]
       --enable-ipv6-ndp                                           Enable IPv6 NDP support
       --enable-k8s                                                Enable the k8s clientset (default true)
       --enable-k8s-api-discovery                                  Enable discovery of Kubernetes API groups and resources with the discovery API
+      --enable-k8s-host-firewall-bypass                           Enable bypassing host firewall for Kubernetes API server access (default true)
```
Small thing, we mention this problem in the host-firewall docs here: https://docs.cilium.io/en/latest/security/host-firewall/#enable-policy-audit-mode-for-the-host-endpoint.  Maybe we should mention this flag there now as well?

joestringer replied:
There's some related discussion here: https://github.com/cilium/cilium/pull/40346#discussion_r2211209892 - My gut feeling was that we're going to want to always enable this flag and probably remove it soon, so it doesn't make sense to document widely and teach users about the flag. 

I'm not sure concretely if there's some other change we should improve in that docs page relating to this feature. At a glance I don't see anything particular to kube-apiserver connectivity in the text there.

tommyp1ckles replied:
@joestringer If this is going to be enabled by default, do we still need the mention about users avoiding breaking connections to kube-apiserver?

joestringer replied:
@tommyp1ckles This change only causes connections from the cilium-agent to be allowed; for the rest of the host, it's still subject to host firewall. So I think that note is still valid.

tommyp1ckles commented on daemon/cmd/datapath.go:
```diff
@@ -367,3 +372,35 @@ func setupRouteToVtepCidr(logger *slog.Logger) error {
+func (*k8sHostFirewallBypass) ConfigureK8sClientsetDialer(dialer *net.Dialer) {
+	dialer.Control = setProxyEgressMark
+	dialer.Resolver = &net.Resolver{
+		PreferGo: true,
```
question: was this the case prior to this change?  If not, what are the implications of having this be true?

atykhyy replied:
It was. [Go documentation](https://pkg.go.dev/net#hdr-Name_Resolution) says that on Unix, the pure go resolver is preferred. I put this in anyway to express intent (setting the resolver's dialer makes no sense unless the pure go resolver is used) and also in case somebody runs the daemon with `GODEBUG=netdns=cgo` or similar.

tommyp1ckles replied:
Ok, thanks for the clarification.

tommyp1ckles commented on pkg/k8s/hostfirewallbypass/bypass.go:
```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: Apache-2.0
```
minor thing, we usually define the `var Cell = ...` and often the config type in a file called `<mypackage>/cell.go`. But this is just a convention, not a rule, that makes looking for the hive related code easier to find.
